src/PureSMT/Process.hs

```diff
@@ -39,26 +42,44 @@ initZ3Instance dbg = do
                      Z3_del_config(cfg);
                      return ctx;
                      } |]
-  let s = Solver solverCtx dbg
+  solverQueue <- newIORef mempty
+  let s = Solver solverCtx dbg solverQueue
   setOption s ":print-success" "true"
   setOption s ":produce-models" "true"
 
   return s
 
--- | Have Z3 evaluate a command in SExpr format.
+-- | Send a bytestring to Z3.
 -- This function is thread-safe as long as concurrent instances do not share the
 -- same logical context.
-command :: Solver -> SExpr -> IO SExpr
-command solver cmd = do
-  let cmdTxt = serializeSExpr cmd
-  when (debugMode solver) $ do
-    BS.putStrLn $ "[send] " `BS.append` cmdTxt
+send :: Solver -> BS.ByteString -> IO (BS.ByteString)
+send solver cmd = do
   let ctx = context solver
-  resp <-
-    [CU.exp| const char* {
-                  Z3_eval_smtlib2_string($fptr-ptr:(Z3_context ctx), $bs-ptr:cmdTxt)
-                  } |]
-      >>= BS.packCString
+  when (debugMode solver) $ do
+    BS.putStrLn $ "[send] " `BS.append` cmd
+  [CU.exp| const char* {
+         Z3_eval_smtlib2_string($fptr-ptr:(Z3_context ctx), $bs-ptr:cmd)
+         } |]
+    >>= BS.packCString
+
+-- | Force the solver to evaluate the queued commands.
+-- The result is not checked to ensure the commands were evaluated correctly.
+flush :: Solver -> IO ()
+flush solver = do
+  cmds <- do
+    let solverQueue = queue solver
+    cmds <- readIORef solverQueue
+    writeIORef solverQueue mempty
+    return $ serializeBatch cmds
+  _ <- send solver cmds
+  return ()
+
+-- | Have Z3 evaluate a command in SExpr format.
+command :: Solver -> SExpr -> IO SExpr
+command solver expr = do
+  flush solver
+  let cmd = serializeSingle $ renderSExpr expr
+  resp <- send solver cmd
```

Comment:
Because I expect the queue to produce an output that is `(success)` x n which makes it harder to parse the result of the last command, which we're actually interested in.

Reply:
Ah, makes sense. Maybe you could mentioned that in a comment.

Reply:
I think I'll just disable the `:print-success` option when not in debug mode.